vtkext/private/module/vtkF3DMetaImporter.cxx

```diff
@@ -0,0 +1,712 @@
+  auto& names =
+    useCellData ? this->Pimpl->CellDataArrayNames : this->Pimpl->PointDataArrayNames;
+
+  if (index < 0 || index >= static_cast<int>(data.size()))
+  {
+    return false;
+  }
+
+  info = data[names[index]];
+  return true;
+}
+
+//----------------------------------------------------------------------------
+int vtkF3DMetaImporter::GetNumberOfIndexesForColoring(bool useCellData)
+{
+  if (!this->Pimpl->ColoringInfoUpdated)
+  {
+    this->UpdateInfoForColoring();
+  }
+
+  auto& data =
+    useCellData ? this->Pimpl->CellDataColoringInfo : this->Pimpl->PointDataColoringInfo;
+  return static_cast<int>(data.size());
+}
+
+//----------------------------------------------------------------------------
+int vtkF3DMetaImporter::FindIndexForColoring(bool useCellData, const std::string& arrayName)
+{
+  if (!this->Pimpl->ColoringInfoUpdated)
+  {
+    this->UpdateInfoForColoring();
+  }
+
+  auto& data =
+    useCellData ? this->Pimpl->CellDataColoringInfo : this->Pimpl->PointDataColoringInfo;
+
+  auto it = data.find(arrayName);
+  if (it != data.end())
+  {
+    return it->second.Index;
+  }
+  return -1;
+}
+
+//----------------------------------------------------------------------------
+const std::vector<vtkF3DMetaImporter::ColoringStruct>&
+vtkF3DMetaImporter::GetColoringActorsAndMappers()
+{
+  return this->Pimpl->ColoringActorsAndMappers;
+}
+
+//----------------------------------------------------------------------------
+const std::vector<vtkF3DMetaImporter::PointSpritesStruct>&
+vtkF3DMetaImporter::GetPointSpritesActorsAndMappers()
+{
+  return this->Pimpl->PointSpritesActorsAndMappers;
+}
+
+//----------------------------------------------------------------------------
+const std::vector<vtkF3DMetaImporter::VolumeStruct>& vtkF3DMetaImporter::GetVolumePropsAndMappers()
+{
+  return this->Pimpl->VolumePropsAndMappers;
+}
+
+//----------------------------------------------------------------------------
+bool vtkF3DMetaImporter::Update()
+{
+  assert(this->RenderWindow);
+  this->Renderer = this->RenderWindow->GetRenderers()->GetFirstRenderer();
+  assert(this->Renderer);
+
+  // XXX: Doing this means that coloring information
+  // is recomputed from scratch when doing incremental loading
+  // This is the simplest way to implement coloring
+  // and should not have too big of an overhead.
+  this->Pimpl->ColoringInfoUpdated = false;
+
+  vtkIdType localCameraIndex = -1;
+
+  if (this->Pimpl->CameraIndex.has_value())
+  {
+    if (this->Pimpl->CameraIndex < 0)
+    {
+      F3DLog::Print(F3DLog::Severity::Warning,
+        "Invalid camera index: " + std::to_string(this->Pimpl->CameraIndex.value()) +
+          ". Camera may be incorrect.");
+    }
+    localCameraIndex = this->Pimpl->CameraIndex.value();
+  }
+
+  for (auto& importerPair : this->Pimpl->Importers)
+  {
+    vtkImporter* importer = importerPair.first;
+
+    // Importer has already been updated
+    if (importerPair.second)
```

Comment:
Let's try to use structured binding when looping on `this->Pimpl->Importers`. `.first`/`.second` are too vague.